File: tfcc_code_generator/ir/node/rnn/bidirectionallstm.py
# Copyright 2021 Wechat Group, Tencent
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
from ir.node import Node
import ir.framework


# ICFO kernel format
class BidirectionalLSTM(Node):
    def update_attributes(self, hidden_size):
        assert isinstance(hidden_size, int) and hidden_size > 0
        self._hidden_size = hidden_size

    def inference(self):
        assert len(self.inputs) == 6
        assert len(self.outputs) == 6
        dtype = self.inputs[0].dtype
        assert all([dtype == inp.dtype for inp in self.inputs])
        assert all([inp.is_tensor() for inp in self.inputs])

        # input_symbol: [seq_length, batch_size, input_size]
        # input_kernel_symbol: [2, input_size, 4 * hidden_size]
        # state_kernel_symbol: [2, hidden_size, 4 * hidden_size]
        # bias_symbol: [2, 4 * hidden_size]
        # initial_h: [2, batch_size, hidden_size]
        # initial_c: [2, batch_size, hidden_size]
        (
            input_symbol,
            input_kernel_symbol,
            state_kernel_symbol,
            bias_symbol,
            initial_h,
            initial_c,
        ) = self.inputs

        assert len(input_symbol.shape) == 3
        seq_length, batch_size, input_size = input_symbol.shape

        # check input kernel
        assert len(input_kernel_symbol.shape) == 3
        assert (
            isinstance(input_kernel_symbol.shape[0], str)
            or input_kernel_symbol.shape[0] == 2
        )
        if isinstance(input_size, str):
            if isinstance(input_kernel_symbol.shape[1], int):
                input_size = input_kernel_symbol.shape[1]
        else:
            assert (
                isinstance(input_kernel_symbol.shape[1], str)
                or input_kernel_symbol.shape[1] == input_size
            )
        assert (
            isinstance(input_kernel_symbol.shape[2], str)
            or input_kernel_symbol.shape[2] == self.hidden_size * 4
        )

        # check state kernel
        assert len(state_kernel_symbol.shape) == 3
        assert (
            isinstance(state_kernel_symbol.shape[0], str)
            or state_kernel_symbol.shape[0] == 2
        )
        assert (
            isinstance(state_kernel_symbol.shape[1], str)
            or state_kernel_symbol.shape[1] == self.hidden_size
        )
        assert (
            isinstance(state_kernel_symbol.shape[2], str)
            or state_kernel_symbol.shape[2] == self.hidden_size * 4
        )

        # check bias
        assert len(bias_symbol.shape) == 2
        assert isinstance(bias_symbol.shape[0], str) or bias_symbol.shape[0] == 2
        assert (
            isinstance(bias_symbol.shape[1], str)
            or bias_symbol.shape[1] == self.hidden_size * 4
        )

        # check initial h
        assert len(initial_h.shape) == 3
        assert initial_h.shape[0] == 2
        if isinstance(batch_size, str):
            if isinstance(initial_h.shape[1], int):
                batch_size = initial_h.shape[1]
        else:
            assert (
                isinstance(initial_h.shape[1], str) or initial_h.shape[1] == batch_size
            )
        assert (
            isinstance(initial_h.shape[2], str)
            or initial_h.shape[2] == self.hidden_size
        )

        # check initial c
        assert len(initial_c.shape) == 3
        assert initial_c.shape[0] == 2
        if isinstance(batch_size, str):
            if isinstance(initial_c.shape[1], int):
                batch_size = initial_c.shape[1]
        else:
            assert (
                isinstance(initial_c.shape[1], str) or initial_c.shape[1] == batch_size
            )
        assert (
            isinstance(initial_c.shape[2], str)
            or initial_c.shape[2] == self.hidden_size
        )

        # forward_output_symbol: [seq_length, batch_size, hidden_size]
        # backword_output_symbol: [seq_length, batch_size, hidden_size]
        # forward_h: [batch_size, hidden_size]
        # backword_h: [batch_size, hidden_size]
        # forward_c: [batch_size, hidden_size]
        # backword_c: [batch_size, hidden_size]
        (
            forward_output_symbol,
            backword_output_symbol,
            forward_h,
            backword_h,
            forward_c,
            backword_c,
        ) = self.outputs
        forward_output_symbol.stype = ir.framework.SymbolType.VARIABLE
        forward_output_symbol.dtype = dtype
        forward_output_symbol.shape = [seq_length, batch_size, self.hidden_size]

        backword_output_symbol.stype = ir.framework.SymbolType.VARIABLE
        backword_output_symbol.dtype = dtype
        backword_output_symbol.shape = [seq_length, batch_size, self.hidden_size]

        forward_h.stype = ir.framework.SymbolType.VARIABLE
        forward_h.dtype = dtype
        forward_h.shape = [batch_size, self.hidden_size]

        backword_h.stype = ir.framework.SymbolType.VARIABLE
        backword_h.dtype = dtype
        backword_h.shape = [batch_size, self.hidden_size]

        forward_c.stype = ir.framework.SymbolType.VARIABLE
        forward_c.dtype = dtype
        forward_c.shape = [batch_size, self.hidden_size]

        backword_c.stype = ir.framework.SymbolType.VARIABLE
        backword_c.dtype = dtype
        backword_c.shape = [batch_size, self.hidden_size]

    @property
    def hidden_size(self):
        return self._hidden_size

    @property
    def attributes(self):
        return {"hidden_size": self.hidden_size}

File: tfcc_runtime/operations/rnn/bidirectionallstm.cpp
// Copyright 2021 Wechat Group, Tencent
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "bidirectionallstm.h"

#include <algorithm>

#include "tfcc_runtime/framework/graph.h"
#include "tfcc_runtime/proto/operations/rnn.pb.h"
#include "tfcc_runtime/utils/commutils.h"

namespace tfcc {
namespace runtime {
namespace rnn {

template <class T>
static const char* _wrapper_bidirectional_lstm(
    const tfcc::Tensor<T>* a, const tfcc::Tensor<T>* inputKernel,
    const tfcc::Tensor<T>* stateKernel, const tfcc::Tensor<T>* bias, const tfcc::Tensor<T>* ih,
    const tfcc::Tensor<T>* ic, tfcc::Variable<T>* forwardOutput, tfcc::Variable<T>* backwardOutput,
    tfcc::Variable<T>* forwardH, tfcc::Variable<T>* backwardH, tfcc::Variable<T>* forwardC,
    tfcc::Variable<T>* backwardC) noexcept {
  try {
    tfcc::View<T> forwardInputKernel(*inputKernel, inputKernel->shape(), 0, 1);
    forwardInputKernel.reshape({inputKernel->shape(1), inputKernel->shape(2)});
    tfcc::View<T> backwardInputKernel(*inputKernel, inputKernel->shape(), 1, 2);
    backwardInputKernel.reshape({inputKernel->shape(1), inputKernel->shape(2)});
    tfcc::View<T> forwardStateKernel(*stateKernel, stateKernel->shape(), 0, 1);
    forwardStateKernel.reshape({stateKernel->shape(1), stateKernel->shape(2)});
    tfcc::View<T> backwardStateKernel(*stateKernel, stateKernel->shape(), 1, 2);
    backwardStateKernel.reshape({stateKernel->shape(1), stateKernel->shape(2)});
    tfcc::View<T> forwardBias(*bias, bias->shape(), 0, 1);
    forwardBias.reshape({bias->shape(1)});
    tfcc::View<T> backwardBias(*bias, bias->shape(), 1, 2);
    backwardBias.reshape({bias->shape(1)});
    tfcc::View<T> forwardIH(*ih, ih->shape(), 0, 1);
    forwardIH.reshape({ih->shape(1), ih->shape(2)});
    tfcc::View<T> backwardIH(*ih, ih->shape(), 1, 2);
    backwardIH.reshape({ih->shape(1), ih->shape(2)});
    tfcc::View<T> forwardIC(*ic, ic->shape(), 0, 1);
    forwardIC.reshape({ic->shape(1), ic->shape(2)});
    tfcc::View<T> backwardIC(*ic, ic->shape(), 1, 2);
    backwardIC.reshape({ic->shape(1), ic->shape(2)});

    std::tie(*forwardOutput, *forwardH, *forwardC) = tfcc::rnn::lstm_forward(
        *a, forwardInputKernel, forwardStateKernel, forwardBias, forwardIH, forwardIC);
    std::tie(*backwardOutput, *backwardH, *backwardC) = tfcc::rnn::lstm_backward(
        *a, backwardInputKernel, backwardStateKernel, backwardBias, backwardIH, backwardIC);
  } catch (std::exception& e) {
    thread_local std::string reason = e.what();
    return reason.c_str();
  }
  return nullptr;
}

template <tfcc::runtime::common::DataType DTYPE>
std::string BidirectionalLSTM<DTYPE>::getOperationTypeUrl() const {
  tfcc::runtime::operations::rnn::BidirectionalLSTM operation;
  return get_protobuf_type_url(operation);
}

template <tfcc::runtime::common::DataType DTYPE>
std::set<unsigned> BidirectionalLSTM<DTYPE>::getOperationVersions() const {
  std::set<unsigned> versions = {tfcc::runtime::operations::rnn::BidirectionalLSTM::VERSION_1};
  return versions;
}

template <tfcc::runtime::common::DataType DTYPE>
std::unique_ptr<OperationResource> BidirectionalLSTM<DTYPE>::process(
    const tfcc::runtime::model::Node& node, Graph& graph) const {
  if (node.inputs_size() != 6 || node.outputs_size() != 6) {
    return nullptr;
  }
  if (!node.operation().Is<tfcc::runtime::operations::rnn::BidirectionalLSTM>()) {
    return nullptr;
  }
  auto& symbolManager = getSymbolManager(graph);
  if (symbolManager.getSymbolInfo(node.outputs(0)).dtype != DTYPE) {
    return nullptr;
  }

  std::unique_ptr<OperationResource> resource(new OperationResource);
  tfcc::runtime::operations::rnn::BidirectionalLSTM operation;
  node.operation().UnpackTo(&operation);

  if (operation.hidden_size() == 0) {
    return nullptr;
  }

  auto& jit = getJIT(graph);

  const tfcc::Tensor<T>* inputSymbol = symbolManager.getTensor(node.inputs(0), T());
  const tfcc::Tensor<T>* inputKernelSymbol = symbolManager.getTensor(node.inputs(1), T());
  const tfcc::Tensor<T>* stateKernelSymbol = symbolManager.getTensor(node.inputs(2), T());
  const tfcc::Tensor<T>* biasSymbol = symbolManager.getTensor(node.inputs(3), T());
  const tfcc::Tensor<T>* initialHSymbol = symbolManager.getTensor(node.inputs(4), T());
  const tfcc::Tensor<T>* initialCSymbol = symbolManager.getTensor(node.inputs(5), T());
  tfcc::Variable<T>* forwardOutputSymbol = symbolManager.getVariable(node.outputs(0), T());
  tfcc::Variable<T>* backwardOutputSymbol = symbolManager.getVariable(node.outputs(1), T());
  tfcc::Variable<T>* forwardHSymbol = symbolManager.getVariable(node.outputs(2), T());
  tfcc::Variable<T>* backwardHSymbol = symbolManager.getVariable(node.outputs(3), T());
  tfcc::Variable<T>* forwardCSymbol = symbolManager.getVariable(node.outputs(4), T());
  tfcc::Variable<T>* backwardCSymbol = symbolManager.getVariable(node.outputs(5), T());

  callFunction(
      jit, _wrapper_bidirectional_lstm<T>, inputSymbol, inputKernelSymbol, stateKernelSymbol,
      biasSymbol, initialHSymbol, initialCSymbol, forwardOutputSymbol, backwardOutputSymbol,
      forwardHSymbol, backwardHSymbol, forwardCSymbol, backwardCSymbol);
  return resource;
}

std::vector<std::unique_ptr<Operation>> get_bidirectional_lstm_operations() {
#define DEFINE_FUNC(dtype) \
  operations.emplace_back(std::unique_ptr<Operation>(new BidirectionalLSTM<dtype>()));
  std::vector<std::unique_ptr<Operation>> operations;
  TFCC_RUNTIME_FOR_ALL_DATA_TYPE(DEFINE_FUNC);
  return operations;
}

}  // namespace rnn
}  // namespace runtime
}  // namespace tfcc
